### ISSUE TITLE
Possible race condition ?

### ISSUE THREAD

**Issue by boxerab** [taskflow/taskflow#397]
Running my program with `helgrind` and `drd`, this taskflow code was flagged, as the condition variable notify
is not protected by mutex. It look like this is by design, but is it not necessary to have both wait and notify
protected by mutex ?

``
  void _park(Waiter* w) {
    std::unique_lock<std::mutex> lock(w->mu);
    while (w->state != Waiter::kSignaled) {
      w->state = Waiter::kWaiting;
      w->cv.wait(lock);
    }
  }

  void _unpark(Waiter* waiters) {
    Waiter* next = nullptr;
    for (Waiter* w = waiters; w; w = next) {
      next = w->next.load(std::memory_order_relaxed);
      unsigned state;
      {
        std::unique_lock<std::mutex> lock(w->mu);
        state = w->state;
        w->state = Waiter::kSignaled;
      }
      // Avoid notifying if it wasn't waiting.
      if (state == Waiter::kWaiting) w->cv.notify_one();
    }
  }
```

**Comment by tsung-wei-huang** [comment]
Yes, you can avoid that warning - this notification doesn't need to be exact so no need to protect them under the lock. I have not yet seen any race report from `tsan`.

BTW, would you mind giving me your email :)? You can shoot me an email at tsung-wei.huang@utah.edu. I just have a question for you about your experience of using taskflow in your project.
